cxx/gendb.cc

```diff
@@ -350,24 +365,385 @@ void GenDB::incorporate_reference_relation(
+// transition_reference. This storage isn't strictly necessary, since we could
+// use reference_values to deduce the values that were unincorporated from the
+// entity CRPs, but this is a bit more convenient. For the top-level call (from
+// transition_reference, is_ancestor_reference=false), we do not accumulate logp
+// or unincorporate from the entity CRP, since these are handled separately for
+// the reference that's being transitioned.
+double GenDB::unincorporate_from_entity_cluster(
+    const std::string& class_name, const std::string& ref_field,
+    const int class_item,
+    std::map<std::tuple<std::string, std::string, int>, int>& unincorporated,
+    const bool is_ancestor_reference) {
+  double logp_adj = 0.;
+
+  int ref_id = get_reference_id(class_name, ref_field, class_item);
+  int ref_item = reference_values.at({class_name, ref_field, class_item});
+
+  const std::string& ref_class =
+      std::get<ClassVar>(schema.classes.at(class_name).vars.at(ref_field).spec)
+          .class_name;
+  CRP& crp = domain_crps.at(ref_class);
+  if (is_ancestor_reference) {
+    crp.unincorporate(ref_id);
+    unincorporated[{class_name, ref_field, class_item}] = ref_item;
+  }
+
+  if (crp.tables.contains(ref_item)) {
+    if (is_ancestor_reference) {
+      logp_adj += crp.logp(ref_item);
+    }
+
+    for (const auto& [name, var] : schema.classes.at(ref_class).vars) {
+      if (std::holds_alternative<ClassVar>(var.spec)) {
+        // If this reference value was the only one in its entity_crp table,
+        // then the entity is removed from its class and we need to
+        // recursively unincorporate *its* reference values.
+        logp_adj += unincorporate_from_entity_cluster(ref_class, name, ref_item,
+                                                      unincorporated);
+      }
+    }
+  } else {
+    if (is_ancestor_reference) {
+      logp_adj += log(crp.alpha) - log(crp.N + crp.alpha);
     }
   }
+  return logp_adj;
+}
+
+// Recursively unincorporates from base relations and stores values.
+template <typename T>
+double GenDB::unincorporate_reference_relation_singleton(
+    Relation<T>* rel, const std::string& rel_name, const T_items& items,
+    std::map<std::string,
+             std::unordered_map<T_items, ObservationVariant, H_items>>&
+        stored_value_map) {
+  // We can stop unincorporating from base relations if the index of
+  // the domain of interest is greater than the number of domains.
+  const std::vector<std::string>& domains = std::visit(
+      [&](auto tr) { return tr.domains; }, hirm->schema.at(rel_name));
+  if (!rel->get_data().contains(items) ||
+      domain_crps.at(domains.back()).tables.contains(items.back())) {
+    return 0.;
+  }
+
+  // Store the items/value and unincorporate them.
+  T value = rel->get_value(items);
+  stored_value_map[rel_name][items] = value;
+
+  rel->unincorporate_from_cluster(items);
+  std::mt19937* prng = nullptr;  // unused
```

Comment:
It's important because this method needs to be entirely deterministic, as it is reincorporating previously-sampled values and not sampling any new ones.